### ISSUE TITLE
exportjson export problems in allegro 23.1

### ISSUE THREAD

**Issue by Jinying-Tang** [juulsA/exportJson#14]
1、it can't export boardGeometry\silkscreen_top(silkscreen_bottom)  with text element 
2、the exported silkscreen layer with solid shape only display the  shape's outline
3、the exported assemble layer with shape doesn‘t display

**Comment by juulsA** [comment]
Hi, 
thanks for reporting these errors ... if can replicate them, but it will take some time to fix them ... 
Regards

**Comment by Jinying-Tang** [comment]
> Hi, thanks for reporting these errors ... if can replicate them, but it will take some time to fix them ... if can replicate them, but it will take some time to fix them ... Thanks for reporting these errors ...... if can replicate them, but it will take some time to fix them ... ...... Regards

thanks for reply
Attached is a pcb that can produce these errors, designed using allegro 23.1
[board.zip](https://github.com/user-attachments/files/15574576/board.zip)

**Comment by juulsA** [comment]
Thanks for sharing the board file, but I'm still using 22.1 ... so I can't open the file. But please check my last commit, most of the errors should be resolved ... for the problem with the unfilled shape, I added a fix, if the shape has no void, otherwise only the outline is drawn ... I have to look into that deeper ... please let me know, what errors are fixed now.

**Comment by Jinying-Tang** [comment]
I've used the version you last submitted
Attached is a pcb that can produce these errors, designed downrev to allegro 17.4
[board_174.zip](https://github.com/user-attachments/files/15579868/board_174.zip)

**Comment by juulsA** [comment]
Thanks! I've done the export an ibom generation, but it looks good for me ... can you tell me what exactly you're missing and could you also provide the `config.json` you're using?

**Comment by Jinying-Tang** [comment]
> Thanks! I've done the export an ibom generation, but it looks good for me ... can you tell me what exactly you're missing and could you also provide the `config.json` you're using?

I didn't specify a config.json file, so I used the default config
![image](https://github.com/juulsA/exportJson/assets/131565469/4f4280d9-3603-45e8-a7b2-6d155741476a)
you can see a text JY-022 in boardGeometry\silkscreen_bottom in .brd file,but in ibom do not see that text
you can see a solid triangle  in boardGeometry\silkscreen_bottom in .brd file, but in ibom only display a triangle  outline

**Comment by juulsA** [comment]
please check again ... don't know why these lines solve the problem, but it seems they do...

**Comment by Jinying-Tang** [comment]
is your ibom generated by board_174.brd displayed all good？

**Comment by juulsA** [comment]
I'm sorry, can't upload images ... this is what it looks after I created the ibom.
[ibom_22.1.png.zip](https://github.com/user-attachments/files/15587282/ibom_22.1.png.zip)

**Comment by Jinying-Tang** [comment]
> I'm sorry, can't upload images ... this is what it looks after I created the ibom. [ibom_22.1.png.zip](https://github.com/user-attachments/files/15587282/ibom_22.1.png.zip)

your ibom looks like all good! Could it be caused by allegro version issue?

**Comment by juulsA** [comment]
I'm not sure ... could you provide the xxx.json from which the html is generated? Than I can compare both.

**Comment by Jinying-Tang** [comment]
> I'm not sure ... could you provide the xxx.json from which the html is generated? Then I can compare both.

thanks for reply!
i use the last commit, exportjson(6 hours ago), now the ibom looks like all good.
![image](https://github.com/juulsA/exportJson/assets/131565469/a7e3b92a-7c31-4937-b2dc-16b111ce66d9)

but the exported assemble layer with shape doesn‘t display still exist,
![image](https://github.com/juulsA/exportJson/assets/131565469/a2909163-3c94-491f-8a4d-5749a84253c1)
![image](https://github.com/juulsA/exportJson/assets/131565469/7f380823-8f10-4d21-a8cf-d2b2efec40f7)

**Comment by juulsA** [comment]
I meant the xxx.json file generated by the script... if you look in the allegro folder, there is a folder named "json", if you could share the file, it would be helpful for me.

**Comment by Jinying-Tang** [comment]
> I meant the xxx.json file generated by the script... if you look in the allegro folder, there is a folder named "json", if you could share the file, it would be helpful for me.我指的是脚本生成的 xxx.json 文件......如果你查看 allegro 文件夹，有一个名为 "json "的文件夹，如果你能共享该文件，对我将很有帮助。

h
[Uploading board.json…]()
ere is the json file

**Comment by juulsA** [comment]
Behind the link is no json file ... could you upload it again, but with the newest commit?

1. you're right, fixed that
2. same ...
3. don't think this is a problem, but I keep that in mind.

Thanks for your help!

**Comment by Jinying-Tang** [comment]
> Behind the link is no json file ... could you upload it again, but with the newest commit?
> 
> 1. you're right, fixed that
> 2. same ...
> 3. don't think this is a problem, but I keep that in mind.
> 
> Thanks for your help!

thanks,
1、when use default config, allegro display :CDS version: 23.100000 "Export complete!"
When the config file is specified, allegro will display nil after "Export complete!"
sometimes , there will a "t" after "Export complete!", what does "nil" "t" mean?

2 、 exported ibom assemble layer with rectangle doesn‘t display still exist,

**Comment by juulsA** [comment]
This is what I get, when I generate the ibom from your JSON file. 
[ibom.zip](https://github.com/user-attachments/files/15612255/ibom.zip)

Maybe you should think about setting `"textsAsSvgPaths" = true`, because only then the font looks like in allegro, otherwise the newstroke font is used.

1. `nil` and `t` are return types from functions or variables. In skill the last used variable in a function is returned, and this is what you see, so don't mind. 
2. do you use the latest interactivehtlmbom?

**Comment by Jinying-Tang** [comment]
> This is what I get, when I generate the ibom from your JSON file.这是我根据您的 JSON 文件生成 ibom 时得到的结果。 [ibom.zip](https://github.com/user-attachments/files/15612255/ibom.zip)
> 
> Maybe you should think about setting `"textsAsSvgPaths" = true`, because only then the font looks like in allegro, otherwise the newstroke font is used.也许你应该考虑设置 `"textsAsSvgPaths" = true` ，因为只有这样字体才看起来像 allegro 中的字体，否则就会使用 newstroke 字体。
> 
> 1. `nil` and `t` are return types from functions or variables. In skill the last used variable in a function is returned, and this is what you see, so don't mind. `nil` 和 `t` 是函数或变量的返回类型。在技能中，函数中最后使用的变量会被返回，这就是你所看到的，所以不用在意。
> 2. do you use the latest interactivehtlmbom?您使用最新的交互式 htlmbom 吗？

yes, i use the lastest interactivehtmltom, and the default config "textsAsSvgPaths" = true

your generated the ibom , the assemble layer with rectangle doesn‘t display still exist
You can see the differences at the arrows in the two pictures.

![1717663426482](https://github.com/juulsA/exportJson/assets/131565469/7623ef73-969b-47c7-9fec-a1244bb29eb1)
![1717663534449](https://github.com/juulsA/exportJson/assets/131565469/ab3830a9-bc05-454f-8b6b-c2e6d1a23178)

**Comment by juulsA** [comment]
Yes, you're right ... seems like I forgot to implement the type `polygon` ... please check the last commit again ... hopefully it's working now...

**Comment by Jinying-Tang** [comment]
> Yes, you're right ... seems like I forgot to implement the type `polygon` ... please check the last commit again ... hopefully it's working now...

yes, use last commit, it works perfect ! thanks a lot!
Here's a small suggestion:
In some PCB designs, for certain irregular slots, we usually use Ncroute_path and Ncroute_plated. These elements affect the appearance of the PCB.
so, Can these parameters also be exported to the iBOM?

**Comment by juulsA** [comment]
I know what you mean ... I don't think this is impossible, but I'm not really sure I want to implement that ... if something is plated or not can't be displayed in the ibom, so I think the best way to get the closest representation of the pcb, is to draw these slots on the `BOARD GEOMETRY/CUTOUT` layer.

**Comment by Jinying-Tang** [comment]
> I know what you mean ... I don't think this is impossible, but I'm not really sure I want to implement that ... if something is plated or not can't be displayed in the ibom, so I think the best way to get the closest representation of the pcb, is to draw these slots on the `BOARD GEOMETRY/CUTOUT` layer.

if some slots need plated, and some slots are none plated, Drawing these slots on the board geometry/cutout layer cannot be separated.

**Comment by juulsA** [comment]
Yes thats right, I meant, that you can draw them on this layer in addition to the `Ncroute_path` and `Ncroute_plated` layers.

**Comment by Jinying-Tang** [comment]
> Yes thats right, I meant, that you can draw them on this layer in addition to the `Ncroute_path` and `Ncroute_plated` layers, to get a close representation of the pcb for the ibom rendering.

Yes, this is a good idea.
Here's a another small suggestion:
allegro old version , ie 16.6 ,  use boardgeometry /outline for baod shape,  when i use 23.1 to open 16.6 .brd, and use exportjson to export ibom, there is an error occured, the command windows display *Error* No board geometry found ...
I believe that regardless of the version, outline or design outline is considered the board's shape,  and inside the board, outline or cutout is treated as void.

**Comment by juulsA** [comment]
Sry for my late response ... I never had the chance to check that, so this was an approach for users using older files, but I think you're right, the mapping should not be dependent of the version... it's on my list, thanks!

**Comment by Jinying-Tang** [comment]
> Sry for my late response ... I never had the chance to check that, so this was an approach for users using older files, but I think you're right, the mapping should not be dependent of the version... it's on my list, thanks!

Thank you for your update. When converting to ibom, I have another issue: for lower versions of outline and cutout, if they are drawn using line segments or arcs (not shape types), these elements do not appear in the exported ibom.
